scylla/src/transport/load_balancing/default.rs

```diff
@@ -125,7 +150,25 @@ or refrain from preferring datacenters (which may ban all other datacenters, if
 
         // If token is available, get a shuffled list of alive replicas.
         let maybe_replicas = if let Some(ts) = &routing_info.token_with_strategy {
-            let local_replicas = self.shuffled_replicas(ts, true, Self::is_alive, cluster);
+            let local_rack_replicas = self.replicas(
+                ts,
+                true,
+                |node| {
+                    Self::is_alive(node)
+                        && self.preferred_rack.is_some()
+                        && node.rack == self.preferred_rack
+                },
+                cluster,
+            );
+            let local_replicas = self.shuffled_replicas(
+                ts,
+                true,
+                |node| {
+                    Self::is_alive(node)
+                        && (self.preferred_rack.is_none() || node.rack != self.preferred_rack)
+                },
+                cluster,
+            );
```

Comment:
`self.replicas`, `self.shuffled_replicas` and `self.pick_replica` had a very similar signature. I feel that it's a bit inconsistent that you introduced a `ReplicaLocation` enum only in `self.pick_replica`. Was there any specific reason behind it?

Reply:
After your changes, semantics of `self.shuffled_replicas` and `self.pick_replica` differ. `self.pick_replica` modifies its predicate according to the value of `ReplicaLocation` argument, but `self.shuffled_replicas` does not. Why?

Reply:
You're right, the predicate logic was passed in the `fallback` function, so I thought it is redundant to modify the predicate in `self.replicas`. I fixed it, but the predicate modification required the `ReplicaLocation` enum to implement `Clone` and `Copy`.